### PR TITLE
Allow changing a clipper's type (ShapeBorderClipper,_BottomAppBarClipper).

### PR DESCRIPTION
**Pull request by amirh**
The downcast was crashing when toggling hasNotch for BottomAppBar.

### CODE REVIEW

File: packages/flutter/test/material/bottom_app_bar_test.dart [hunk 1]
```diff
@@ -86,6 +86,37 @@ void main() {
     expect(physicalShape.color, const Color(0xff0000ff));
   });
 
+  testWidgets('toggle hasNotch', (WidgetTester tester) async {
+    await tester.pumpWidget(
```
Comment (HansMuller) [inline packages/flutter/test/material/bottom_app_bar_test.dart]: A comment explaining why we're toggling hasNotch...

**Comment by HansMuller** [comment]
LGTM

File: packages/flutter/test/material/bottom_app_bar_test.dart (hunk 1 above)
Comment (amirh) [reply packages/flutter/test/material/bottom_app_bar_test.dart]: done

File: packages/flutter/lib/src/material/bottom_app_bar.dart [hunk 2]
```diff
@@ -152,7 +152,11 @@ class _BottomAppBarClipper extends CustomClipper<Path> {
   }
 
   @override
-  bool shouldReclip(covariant _BottomAppBarClipper oldClipper) {
-    return oldClipper.geometry != geometry;
+  bool shouldReclip(CustomClipper<Path> oldClipper) {
+    if (oldClipper.runtimeType != _BottomAppBarClipper)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/bottom_app_bar.dart]: this is impossible. oldClipped is guaranteed to have the same runtimeType as `this`
(unless there's a bug in the code that calls `shouldReclip`, which is possible)

**Comment by Hixie** [comment]
I think this fix is incorrect. The `shouldRefoo(oldThing)` pattern is supposed to guarantee that `oldThing` is the same type, otherwise it's guaranteed that you should refoo.

**Comment by Hixie** [comment]
cc @amirh see comment above

**Comment by amirh** [comment]
Looks like there was indeed a bug in _RenderCustomClip, sent a fix in #18248.

**Pull request merged**